test/kafka-upsert-benchmark/mzworkflows.py

```diff
@@ -0,0 +1,165 @@
+        command=" ".join(
+            [
+                f"--num-records={num_records}",
+                "--keys=random",
+                "--key-min=0",
+                f"--key-max={num_keys}",
+                "--values=bytes",
+                f"--max-message-size={value_bytes+1}",
+                f"--min-message-size={value_bytes}",
+                "--topic=testdrive-load-test-1",
+                "--quiet",
+            ]
+        ),
+    )
+
+    if restart_mz:
+        w.start_services(services=[mz_service])
+        w.wait_for_mz(service=mz_service)
+    expected = records_sent + num_records
+
+    host_port = int(w.composition.find_host_ports(mz_service)[0])
+    query = "SELECT * FROM num_records_ingested"
+    ui.progress(f"waiting for materialize to handle {query!r}", "C")
+    error = None
+    start_time = time.monotonic()
+
+    # Default to polling Materialize once a second because that is the current default
+    # timestamp frequency and polling any more frequently doesn't give new data.
+    for remaining in ui.timeout_loop(timeout_secs):
+        try:
+            conn = pg8000.connect(
+                host="localhost", port=host_port, user="materialize", timeout=1
+            )
+            conn.autocommit = True
+
+            cursor = conn.cursor()
+            cursor.execute("SELECT * FROM num_records_ingested")
+            n = cursor.fetchone()[0]
+            curr_time = time.monotonic()
+            elapsed = curr_time - start_time
+            if n == expected:
+                say(f"query result: {n} after {elapsed}")
+                ui.progress("success!", finish=True)
+                return
+            else:
+                say(f"materialize has not ingested {expected} records, got: {n}")
+        except Exception as e:
+            ui.progress(" " + str(int(remaining)))
+            error = e
+    ui.progress(finish=True)
+    raise UIError(f"never got correct result {expected}: {error}")
+
+
+def workflow_kafka_upsert_benchmark(w: Workflow, args: List[str]):
+    parser = WorkflowArgumentParser(w)
+    parser.add_argument("--num-steps", type=int, default=10)
+    parser.add_argument("--records-per-step", type=int, default=50000)
+    parser.add_argument("--num-keys", type=int, default=1000000)
+    parser.add_argument("--value-bytes", type=int, default=500)
+    parser.add_argument("--initial-records", type=int, default=200000)
+    parser.add_argument("--timeout-secs", type=int, default=120)
+    parser.add_argument("--restart-mz", action="store_true")
+    parser.add_argument("--enable-persistence", action="store_true")
+    args = parser.parse_args(args)
+
+    w.start_and_wait_for_tcp(services=prerequisites)
+    mz_service = "materialized"
+    if args.enable_persistence:
+        mz_service = "materialized_persist"
+
+    w.start_services(services=[mz_service])
+    w.wait_for_mz(service=mz_service)
+
+    w.run_service(
+        service="testdrive-svc",
+        command="setup.td",
```

Comment:
Unfortunately creating the topic/source only once and reusing it by just adding more data will cause the numbers reported to be inaccurate. This is because the upsert operator has a sub-linear performance -- the more data is there, the more time it takes to ingest more of it. For example, with this command line:

```
./mzcompose run  kafka-upsert-benchmark --num-steps 1000 --records-per-step 1000000 --num-keys 1000000000
```

I get those measurements:

```
C> query result: 200000 after 1.0029516108334064
C> query result: 1200000 after 5.0061679147183895
C> query result: 2200000 after 6.006960421800613
C> query result: 3200000 after 7.005936376750469
C> query result: 4200000 after 8.006759010255337
C> query result: 5200000 after 9.04199194535613
C> query result: 6200000 after 9.007915418595076
C> query result: 7200000 after 10.008318159729242
C> query result: 8200000 after 11.008377231657505
C> query result: 9200000 after 10.006969027221203
C> query result: 10200000 after 10.01085114479065
C> query result: 11200000 after 11.007687475532293
C> query result: 12200000 after 10.00821939483285
C> query result: 13200000 after 11.009276680648327
C> query result: 14200000 after 10.007657367736101
C> query result: 15200000 after 11.023996021598577
```

In other words, performance degrades between runs as the total number of keys ingested approaches ```--num-keys```, before stabilizing.

Reply:
I've always heard that described as super-linear performance but I agree! I didn't intend for these measurements to be seen as the canonical measure for how long it takes to ingest in this case 1 MM records, but rather how long it took to get 1 MM additional records from the data we have, to see how ingest perf evolves over time. Put a different way, if we inserted 152 MM records straight up i think it would take something proportional to the sum of the observed times I think.

I unfortunately didn't do the experiment because I totally reworked the benchmark sorry.